### PR TITLE
Add token auth for jwt

### PR DESCRIPTION
**Pull request by sfentress**
@scytacki or @dougmartin Do you want to take a look at the PR?

It adds an optional 3-minute authentication token to an external activity's url, and if an external site calls the JWT API route with the token in the header, it will find the appropriate user and then return a JWT for that user.

This uses a similar pattern as the token for the external reports.

I'm not sure what the CORS settings are for the portal right now. This would be expecting that we can make a GET request to `/api/v1/jwt/firebase`.

### CODE REVIEW

File: app/controllers/api/v1/jwt_controller.rb [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
+      token = $1
+      grant = AccessGrant.find_by_access_token(token)
+
+      return error('Cannot find AccessGrant for token #{token}') if !grant
+      return error('AccessGrant has expired') if grant.access_token_expires_at < Time.now
+
+      user = grant.user
+    else
+      return error('You must be logged in to use this endpoint') if current_visitor.anonymous?
+      user = current_visitor
```
Comment (dougmartin) [inline app/controllers/api/v1/jwt_controller.rb]: I think @scytacki doesn't want us to use current_visitor anymore

File: app/models/external_activity.rb [hunk 2]
```diff
@@ -151,6 +151,10 @@ def url(learner = nil)
       if learner
         append_query(uri, "learner=#{learner.id}") if append_learner_id_to_url
         append_query(uri, "c=#{learner.user.id}") if append_survey_monkey_uid
+        if append_auth_token
+          token = learner.user.create_access_token_valid_for(3.minutes)
+          append_query(uri, "token=#{token}")
+        end
```
Comment (dougmartin) [inline app/models/external_activity.rb]: Not related to the code but do we have a cron job that deletes expired tokens?

Comment (scytacki) [reply app/models/external_activity.rb]: I don't believe we do. In other cases like this we've had the creation code check if there are any expired objects and delete them. That way we don't need a cron job, it does slow down the creation a bit, but this probably isn't a problem.

File: app/controllers/api/v1/jwt_controller.rb (hunk 1 above)
Comment (scytacki) [reply app/controllers/api/v1/jwt_controller.rb]: Yes I'm trying to phase out current_visitor.  current_user is what you should use.  current_user will be nil if the is no user logged in.  current_visitor will always be set to something.

**Comment by scytacki** [comment]
Here is where cors is configured: http://github.com/concord-consortium/rigse/blob/9f7d6bd1d431565bbb2d70227ed66ad0f610d3e1/config/application.rb#L119

So unless the portal is configured with the 'allow_cors' portal feature (line 139), then I think this api/v1/jwt request will fail.

File: app/controllers/api/v1/jwt_controller.rb (hunk 1 above)
Comment (sfentress) [reply app/controllers/api/v1/jwt_controller.rb]: :+1: Ok, that's changed. I was having trouble with current_user before, but that actually had to do with how I was passing credentials for a CORS request.

File: app/models/external_activity.rb (hunk 2 above)
Comment (sfentress) [reply app/models/external_activity.rb]: I've now added `AccessGrant.prune!` before the token creation here, matching the pattern used in the AuthController.

**Comment by sfentress** [comment]
@scytacki Do we not still have `allow_cors` set for the GeniGames login? (I'm not sure how to check env settings on production.)

**Comment by scytacki** [comment]
If you are logged into aws you can see the staging env here: https://console.aws.amazon.com/ecs/home?region=us-east-1#/taskDefinitions/learn-ecs-staging-App/42 
Click on the 'App' container, and that will show the environment variables.

On production you can look here:
https://console.aws.amazon.com/ecs/home?region=us-east-1#/taskDefinitions/learn-ecs-production-App/7

So on both of those 'allow_cors' is enabled.  No other ENV are set.
So I think that means anything from *.concord.org will be allowed to access anything.
That would mean any activities on github.io will not be able to get JWT tokens. 
So I think it is best add it to the explicit listing on line 119

**Comment by sfentress** [comment]
The GV report will be at a *.concord.org url once I deploy it to S3, and GV itself is already there, so I don't know if we need to add the explicit whitelist, at least not for this.

(Although I note that the API that the external reports uses with the bearer token already permits requests from outside *.concord.org, and then checks the host explicitly. Is there some reason why a get request to `/api/v1/jwt/firebase` would be blocked by CORS settings while a similar request to  `/api/v1/offerings/` doesn't?)

**Comment by scytacki** [comment]
Because of this line: https://github.com/concord-consortium/rigse/blob/9f7d6bd1d431565bbb2d70227ed66ad0f610d3e1/config/application.rb#L126

It is nice to support github pages, so I think it is best to add a new line here for this route.

**Comment by sfentress** [comment]
:+1: Ok, opened up in the same way.

**Pull request merged**